### PR TITLE
refactor: Remove dependency on SSC from search and taint engines

### PR DESCRIPTION
**Pull request by IagoAbal**
The engines should know nothing about SSC.

test plan:
make test

### CODE REVIEW

**Review by mmcqd — APPROVED** [review]
Nice! Having `match_hook` be allowed to return data seems better anyway

File: src/engine/Match_rules.ml [hunk 1]
```diff
@@ -145,6 +145,22 @@ let per_rule_boilerplate_fn ~timeout ~timeout_threshold =
...
+  |> List.concat_map (fun (pm : Pattern_match.t) ->
+         let rule_id = pm.rule_id.id in
+         let dependency_matches = get_dep_matches rule_id in
+         let pms' =
+           Match_dependency.annotate_pattern_match dependency_matches pm
+         in
+         pms'
+         |> List.iter (fun pm' ->
+                let str =
+                  Common.spf "with rule %s" (Rule_ID.to_string rule_id)
```
Comment (mmcqd) [inline src/engine/Match_rules.ml]: Actually, this bit isn't SSC specific, maybe it should get it's own name?

Comment (IagoAbal) [reply src/engine/Match_rules.ml]: Yeah fair enough, moved that into `print_match`.

File: src/engine/Match_tainting_mode.ml [hunk 2]
```diff
@@ -914,8 +914,8 @@ let check_fundef lang options taint_config opt_ent ctx ?glob_env
   in
   (flow, mapping)
 
-let check_rule ?dep_matches per_file_formula_cache (rule : R.taint_rule)
-    match_hook (xconf : Match_env.xconfig) (xtarget : Xtarget.t) =
+let check_rule per_file_formula_cache (rule : R.taint_rule) match_hook
```
Comment (aryx) [inline src/engine/Match_tainting_mode.ml:917]: We should not call it match_hook but matches_hook then.
I like that we moved SSC code out of here, but I don't like we piggypack on the
match_hook. Maybe it should be a separate hook, or maybe we should be able to do that 
in the caller.

Comment (aryx) [reply src/engine/Match_tainting_mode.ml:917]: The original intent of match_hook was just give control to code up in the call stack to do something with a match, such as printing it the match, incrementally, without having to wait for the whole result.

**Pull request merged**